### PR TITLE
[MKL-DNN] Added reusing of primitive descriptors (fp32)

### PR DESCRIPTION
**Pull request by jczaja**
Changes discussed here are are introducing Reusing concept for MKL-DNN primitive descriptors (computational ones) for fp32 mkl-dnn ops (conv, conv transpose, softmax and batch norm). 

Motivation:
- To be independent from creation time of primitive descriptors, that may vary from version of MKL-DNN.
- Performance improvement

Performance improvements (SKX8180, ResNet50):
- Significant boost for small workload when lots of computing  resources ( BS=1, num threads=28) : ~10% 
- Tiny improvement for more often use scenarios (BS=1, num_threads=1) : ~0.5%

### CODE REVIEW

**Comment by jczaja** [comment]
@luotao1 , @tensor-tang , @kbinias Could you please review this PR?

**Review by kbinias — DISMISSED** [review]
LGTM

**Comment by jczaja** [comment]
@luotao1 , @tensor-tang Do you need any more information on that PR ?

**Comment by tensor-tang** [comment]
Thanks @jczaja 

What it the background of this PR?

Is this the final solution of replacing tensor modifications discussed in #16248 ?

File: paddle/fluid/operators/mkldnn/conv_transpose_mkldnn_op.cc [hunk 1]
```diff
@@ -12,6 +12,7 @@
    See the License for the specific language governing permissions and
    limitations under the License. */
 
+#include <boost/optional.hpp>
```
Comment (tensor-tang) [inline paddle/fluid/operators/mkldnn/conv_transpose_mkldnn_op.cc]: prefer use `""` since boost is in third-party.

**Comment by jczaja** [comment]
@tensor-tang This is not directly related to tensor modifications. It is just optimization. As for tensor modifications, We are looking to have it updated/reimplemented to match mkl-dnn 1.0 new API.

MKL-DNN Primitive descriptor creation time takes quite some time and it varies from MKL-DNN version to another. This change makes PD to be stored in context and reuse in next iterations.

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 2]
```diff
@@ -395,9 +396,28 @@ class TransposeMKLDNNHandler : public MKLDNNHandler {
   std::vector<int> logical_axis_;
 };
 
+template <typename T>
+struct mode;
```
Comment (tensor-tang) [inline paddle/fluid/platform/mkldnn_reuse.h]: what is mode? seems like `mkldnn::algorithm`

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 3]
```diff
@@ -542,6 +562,73 @@ class ConvMKLDNNTemplateHandler : public MKLDNNHandler {
...
+
+    auto conv_pd = std::static_pointer_cast<typename forward_t::primitive_desc>(
+        dev_ctx_.GetBlob(key_conv_pd));
+
+    if (conv_pd == nullptr) {
+      mkldnn::memory::dims stride_dims = strides;
+      mkldnn::memory::dims padding_dims = paddings;
+
+      auto conv_desc =
+          bias ? typename forward_t::desc(fwd_prop_kind, mode<forward_t>::T,
```
Comment (tensor-tang) [inline paddle/fluid/platform/mkldnn_reuse.h]: And it is just for conv now, just use what is needed seems clearer.

**Comment by tensor-tang** [comment]
> This is not directly related to tensor modifications. It is just optimization. As for tensor modifications, We are looking to have it updated/reimplemented to match mkl-dnn 1.0 new API.

Got it~

File: paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc [hunk 4]
```diff
@@ -633,31 +630,6 @@ class ConvMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
   }
 
  private:
-  mkldnn::primitive_attr CreatePostOps(bool fuse_relu,
```
Comment (tensor-tang) [inline paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc:636]: Seems just move this to `mkldnn_reuse.h`?

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 5]
```diff
@@ -542,6 +562,73 @@ class ConvMKLDNNTemplateHandler : public MKLDNNHandler {
...
+      constexpr float placeholder = 0.0f;
+      post_operations.append_eltwise(scale, mkldnn::algorithm::eltwise_relu,
+                                     negative_slope, placeholder);
+    }
+    conv_attr.set_post_ops(post_operations);
+    return conv_attr;
+  }
+
+  std::shared_ptr<typename forward_t::primitive_desc>
+  AcquireConvolutionPrimitiveDescriptor(
```
Comment (tensor-tang) [inline paddle/fluid/platform/mkldnn_reuse.h:591]: Just find that why only `AcquireConvolutionPrimitiveDescriptor` should be place in `mkldnn_reuse.h` while others do not.

File: paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc (hunk 4 above)
Comment (jczaja) [reply paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc:636]: @tensor-tang Yes. Apart from moving CreatePostOps and ConvFwdPrimitiveDesc and ConvTransposeFwdPrimitiveDesc into mkldnn_reuse.h, they were also unified eg.
Two implementations of CreatePostOps were reimplemented into one and two instances of convFwdPrimitiveDesc and two instances ConvTransposeFwdPrimitiveDesc where also replaced with one function:  AcquireConvolutionPrimitiveDescriptor

Comment (tensor-tang) [reply paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc:636]: OK, I see.

So what's the difference with line 661 https://github.com/PaddlePaddle/Paddle/pull/16667/files#diff-ff3a710419e42895ac7587353a7af5bcL661.

Can this be reused there as well ?

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 5 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h:591]: @tensor-tang AcquireConvolutionPrimitiveDescriptor does use CreatePostOps. CreatePostOps where used in conv_mkldnn as well as conv_transpose_mkldnn . I merged two implementations of CreatePostOps into one so it is shared by conv and conv_transpose  and also because it is used to AcquireConvolutionPrimitiveDescriptor , I decided to put it in mkldnn_reuse.h . Do you suggest other location?

Comment (tensor-tang) [reply paddle/fluid/platform/mkldnn_reuse.h:591]: That's fine keeping here, if both conv and transpose conv using these functions.

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 2 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h]: @tensor-tang Yes, mode was introduced so that When AcquireConvolutionPrimitiveDescriptor
is working for conv_mkldnn eg. template parameter is "convolution_forward"  then mkldnn::algorithm should be "convolution_direct" . When AcquireConvolutionPrimitiveDescriptor
is working for conv_transpose_mkldnn eg. template parameter is "deconvolution_forward" then mkldnn::algorithm should be "desconvolution_direct" . So mode was introduced to have this mapping done

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 3 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h]: @tensor-tang AcquireConvolutionPrimitiveDescriptor is used for both conv and conv_transpose.

File: paddle/fluid/operators/mkldnn/conv_transpose_mkldnn_op.cc (hunk 1 above)
Comment (jczaja) [reply paddle/fluid/operators/mkldnn/conv_transpose_mkldnn_op.cc]: ok, will update that

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 3 above)
Comment (tensor-tang) [reply paddle/fluid/platform/mkldnn_reuse.h]: All reference just: https://github.com/PaddlePaddle/Paddle/pull/16667/files#diff-a60a81d3c97ec5a7cf7cf3d7e1aa9adbR608 and https://github.com/PaddlePaddle/Paddle/pull/16667/files#diff-a60a81d3c97ec5a7cf7cf3d7e1aa9adbR612 right?

seems no need add `mode` for these.

File: paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc (hunk 4 above)
Comment (jczaja) [reply paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc:636]: @tensor-tang This is Int8 CreatePostOps it has additional params. In this PR I haven't touched int8 conv. Later on yes we could do similar reusing of functions for conv int8.

Comment (tensor-tang) [reply paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc:636]: Cool.

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 3 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h]: @tensor-tang I do not quite understand. AcquireConvolutionPrimitiveDescriptor is used twice: for conv and conv_transpose. So for conv , template paramter is "convolution_forward" and later on "mode" does set algorithm based on that to "convolution_direct" . For conv_transpose template parameter is "deconvolution_forward" and later on "mode" maps this deconvolution_forward into "deconvolution_direct". So in my opinion it is useful at that scenario.

**Comment by jczaja** [comment]
@tensor-tang  I have updated code to Include boost as third part as suggested.

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 3 above)
Comment (tensor-tang) [reply paddle/fluid/platform/mkldnn_reuse.h]: OK.
Then the name should be adapted, `mode` is not a good name for what is mean. 
`algorithm` should be better, or other names you suggest.

File: paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc [hunk 6]
```diff
@@ -144,7 +144,6 @@ class ConvMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
     const std::string key = platform::ConvMKLDNNHandler::GetHash(
```
Comment (xiaolil1) [inline paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc:144]: GetHash can be replaced by AppendKey, which is more efficient by saving multiple string creation.

File: paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc [hunk 7]
```diff
@@ -191,18 +192,14 @@ class ConvMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
       bias_tz = paddle::framework::vectorize2int(bias->dims());
       auto bias_md = platform::MKLDNNMemDesc(
           bias_tz, platform::MKLDNNGetDataType<T>(), memory::format::x);
-      conv_pd = ConvFwdPrimitiveDesc(
+      conv_pd = handler.AcquireConvolutionPrimitiveDescriptor(
```
Comment (xiaolil1) [inline paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc:195]: It seems these pds (primitive descriptors) still be created in every iterations, right? I have no idea how this PR do "primitive descriptors reusing", could you please explain it? Thanks.

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 8]
```diff
@@ -542,6 +562,73 @@ class ConvMKLDNNTemplateHandler : public MKLDNNHandler {
                                scale_data, mask);
   }
 
+  mkldnn::primitive_attr CreatePostOps(bool fuse_relu,
```
Comment (xiaolil1) [inline paddle/fluid/platform/mkldnn_reuse.h:565]: @jczaja I have no idea about why these code moving can improve the performance, It seems that the creation of primitive descriptors was not decreased as I commented on "conv_mkldnn_op.cc", could you please explain it?

File: paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc (hunk 6 above)
Comment (jczaja) [reply paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc:144]: @xiaolil1 Yes, but someone else from our team is reworking Hash generation so We plan to migrate when he is done

File: paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc (hunk 7 above)
Comment (jczaja) [reply paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc:195]: @xiaolil1  AcquireConvolutionPrimitiveDescriptor internally does create PD or get it from Context eg. reuse it

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 8 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h:565]: @xiaolil1 Just moving this code does not improve performacne. PD reusing improves performance. CreatePostOps was present in conv as well as in conv_transpose, and now instead of two instances we have only one in mkldnn_reuse. So that particular change is refactoring to avoid redundant code. See some comments above

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 3 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h]: I introduced change : "mode" -> "convolutional_algorithm"  . Hopefully it is more descriptive that way.

File: paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc (hunk 7 above)
Comment (xiaolil1) [reply paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc:195]: @jczaja Thanks for your answer. 
I understand pds was reused in  AcquireConvolutionPrimitiveDescriptor, but these "md" like bias_md and "memory" like weights_memory_p, "format", "data_type" ... still were created, which is unnecessary right?

Comment (jczaja) [reply paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc:195]: @xiaolil1  Memory descriptors eg. md  objects are still being created and yes it would be good to remove then later on.

Comment (xiaolil1) [reply paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc:195]: @jczaja OK, you can refer to conv INT8 primitive reuse, which have reused all these "descriptors".

**Comment by jczaja** [comment]
@tensor-tang , @xiaolil1 I have rebased changes to have CI passing. Please continue with review

**Review by kbinias — APPROVED** [review]
LGTM

**Review by xiaolil1 — APPROVED** [review]
LGTM

**Comment by tensor-tang** [comment]
Thanks @jczaja
